test/ha/controller_test.go

```diff
@@ -0,0 +1,69 @@
+// +build e2e
+
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ha
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/e2e"
+)
+
+const (
+	controllerDeploymentName = "controller"
+	controllerLabel          = "app=controller"
+)
+
+func TestControllerHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	if err := scaleUpDeployment(clients, controllerDeploymentName); err != nil {
+		t.Fatalf("Failed to scale deployment: %v", err)
+	}
+	defer scaleDownDeployment(clients, controllerDeploymentName)
+	test.CleanupOnInterrupt(func() { scaleDownDeployment(clients, controllerDeploymentName) })
+
+	service1Names, resources := createPizzaPlanetService(t, "pizzaplanet-service1")
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, service1Names) })
+	defer test.TearDown(clients, service1Names)
+
+	leaderController, err := getLeader(t, clients, controllerDeploymentName, controllerLabel)
+	if err != nil {
+		t.Fatalf("Failed to get leader: %v", err)
+	}
+
+	clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).Delete(leaderController, &metav1.DeleteOptions{})
+
+	if err := waitForPodDeleted(t, clients, leaderController); err != nil {
+		t.Fatalf("Did not observe %s to actually be deleted: %v", leaderController, err)
+	}
+
+	// Make sure a new leader has been elected
+	if _, err = getLeader(t, clients, controllerDeploymentName, controllerLabel); err != nil {
+		t.Fatalf("Failed to find new leader: %v", err)
+	}
```

Comment:
verify here that new leader != old leader (thought I guess it's kind of ensured by the `getLeader` itself)

Reply:
Yeah. In this test we first delete the leader and then we try to get the new leader. And and this point we know for sure that the previous leader was deleted so it MUST be a new leader,  different from the previous one (we check that the new leader pod exists so it really has to be a new leader)

Reply:
Victor is right, as written this test is actually testing effectively nothing.

A `ksvc` can serve just fine without controller pods running, the only pods that actually matter are the autoscaler and the webhook, neither of which are HA right now.

So unless we actually check that something new picks up the lease, this isn't actually testing anything.  I could introduce a chance that breaks leaderelection in a way that additional pods don't pick up the lease and at best this test would become flaky.

Reply:
Hmm. We check that something new pick up the lease. It's ensured by the getLeader function which checks that the pod  that has the lease actually exists. And we previously deleted the pod that had the Lease. So it has to be a new pod that has the lease right now :-/

Reply:
I still think we want a prober running.  There should not be dataplane downtime when you kill one of our controllers.

Reply:
Also the rename when you move the function to PKG will make things considerably clearer.

Reply:
Alright. The function is being renamed and added to PKG in https://github.com/knative/pkg/pull/1259 which is ready for review. The prober test is being added in https://github.com/knative/serving/pull/7777